src/ngamsServer/ngamsServer/ngamsServer.py

```diff
@@ -436,6 +435,68 @@ def proxy_request(self, host_id, host, port, timeout=300):
 
         self.send_data(data, mime_type, code=code, hdrs=hdrs)
 
+    def remote_proxy_request(self, request, host, port, timeout=300):
+        """Proxy the current request to remote host ``host``:``port``"""
+
+        url = 'http://{0}:{1}/RETRIEVE'.format(host, port)
+        logger.info("Proxying request for /RETRIEVE to %s:%d", host, port)
+
+        parameter_list = []
+        for parameter in request.getHttpParNames():
+            if parameter.lower() == "initiator":
+                continue
+            else:
+                parameter_list.append([parameter, request.getHttpPar(parameter)])
+
+        start_byte = 0
+        header_list = []
+        for header in request.getHttpHdrs():
+            header_list.append([header, request.getHttpHdr(header)])
+            if header.lower() == "range":
+                value = request.getHttpHdr(header)
+                start_byte = int(value.replace("bytes=", "").split("-")[0])
+
+        block_size = self.ngasServer.getCfg().getBlockSize()
+
+        # Make sure the time_out parameters is within proper boundaries
+        timeout = min(max(timeout, 0), 1200)
+
+        authorization_header = ngamsSrvUtils.genIntAuthHdr(self.ngasServer)
+
+        response = ngamsHttpUtils.httpGetUrl(url, parameter_list, header_list,
+                                             timeout, authorization_header)
+
+        logger.info("Received remote partner site proxy response from %s:%d, sending to client", host, port)
+        response_headers = {header[0]: header[1] for header in response.getheaders()}
+        logger.info("Headers from remote partner site proxy response: %r", response_headers)
+
+        with contextlib.closing(response):
+            size = int(response.getheader("content-length"))
+            self.write_stream_data(response, response_headers, size, start_byte, block_size)
+
+    def write_stream_data(self, response, headers, size, start_byte=0, block_size=65536):
+        """Streams the data from the remote host"""
+
+        logger.info("Sending %d bytes of data and headers %r", size, headers)
+
+        self.send_response(response.status, hdrs=headers)
+        self.end_headers()
+
+        logger.info("Sending %d bytes to client, starting at byte %d", size, start_byte)
+        data_sent = start_byte
+        data_to_send = size
+        start_time = time.time()
+
+        # TODO: Should we do something about https here?
+        self.wfile.flush()
+        while data_sent < data_to_send:
+            stream_buffer = response.read(block_size)
+            self.wfile.write(stream_buffer)
+            data_sent += len(stream_buffer)
+
+        elapsed_time = time.time() - start_time
+        size_mb = size / 1024. / 1024.
+        logger.info("Sent data stream at %.3f [MB/s]", size_mb / elapsed_time)
```

Comment:
The loop here looks very similar what `write_data` does using `shutil.copyfileobj`, which is basically a read+write loop, maybe it's worth considering merging the two?

HTTPS is indeed something to look at in the future, but not now really; the TODO comment goes really well here.

Also, beware that `elapsed_time` *could* be 0 (chances are low), which causes the division to fail of course. We have seen this in the archiving loop for very small files, in which case we fake a tiny amount of time (would be great to know the precision of the timer and use that instead of a hardcoded 1[us] though):

https://github.com/ICRAR/ngas/blob/c6556bcddf957751dc84556da5417ec4b9fd8fac/src/ngamsServer/ngamsServer/ngamsArchiveUtils.py#L205-L207

Flushing `wfile` shouldn't be necessary in principle, any particular reason for it to be there?

Reply:
I did not know `shutil.copyfileobj()` used a loop. Although it is similar I am not sure they are the same. It does not take into account the start byte offset. And it expects a buffer containing the full data contents (at least I think that is what it is doing). In `write_stream_data()` we stream the data from the response socket. At least, I hope that is what it is doing. Streaming from the socket is more efficient and very important when you are handling large files.

The division by zero is well spotted. I will fix this.

I flush `wfile` to be on the safe side but it is probably unnecessary. I will remove it.

Reply:
OK, I think I understand my confusion now with https://github.com/ICRAR/ngas/pull/28#discussion_r489986466. I had thought that the `start_byte` you calculate in `remote_proxy_request` was then used as an offset when reading data back from the response (i.e., skip the first `start_byte`s from the reponse), while in reality it's used only to ensure that the correct amount of data is read and sent back; the response is still read from the beginning.

It also occurred to me now that if an error happens on the remote host and less data is sent (i.e., small XML error message instead of big file), this loop will run forever because `data_sent` will stop growing and subsequent `read` calls will keep returning an empty string. So the better stop condition really is the end of the stream rather than the number of bytes read.

If I understand correctly, your intent here is to check that in the case of a partial retrieve only that much partial data arrives. I would argue this is unnecessary: in the case of an error the amount of data will be much less, and in the case of a success the amount of data *should be* what you want already; if that was wrong then there would be a more fundamental problem with the implementation of partial retrieves that would impact not only this but also direct requests from users. And if there was a problem. there is really no correction you can do automatically at this point to solve it (e.g., what happens if the unintended extra data was at the beginning of the stream?).

Taking those two points I think it's more useful to use `write_data` here. What *could* be done though is to add a `size` parameter to `write_data` to add the timing information as you did. How does that sound?